File: Vital.sandbox/Sandbox/lua/module/timer.lua
----------------------------------------------------------------
--[[ Resource: Vital.sandbox
     Script: Sandbox: lua: module: timer.lua
     Author: vStudio
     Developer(s): Aviril, Tron, Mario, Аниса, A-Variakojiene
     DOC: 14/09/2022
     Desc: Timer Utilities ]]--
----------------------------------------------------------------


-----------------
--[[ Imports ]]--
-----------------

local imports = {
    type = type,
    tonumber = tonumber,
    coroutine = coroutine,
    math = math
}
coroutine = nil


----------------------
--[[ Class: Timer ]]--
----------------------

local timer = class:create("timer")

function timer.public:create(...)
    local cTimer = self:createInstance()
    if cTimer and not cTimer:load(...) then
        cTimer:destroyInstance()
        return false
    end
    return cTimer
end

function timer.public:destroy(...)
    if not timer.public:isInstance(self) then return false end
    return self:unload(...)
end

function timer.public:load(exec, interval, executions, ...)
    if not timer.public:isInstance(self) then return false end
    interval, executions = imports.tonumber(interval), imports.tonumber(executions)
    if not exec or (imports.type(exec) ~= "function") or not interval or not executions then return false end
    interval, executions = imports.math.max(1, interval), imports.math.max(0, executions)
    self.exec = exec
    self.currentExec = 0
    self.interval, self.executions = interval, executions
    self.arguments = table.pack(...)
    imports.coroutine.resume(imports.coroutine.create(function()
        while ((self.executions == 0) or (self.currentExec < self.executions)) do
            imports.coroutine.sleep(self.interval)
            if not timer.public:isInstance(self) then return false end
            self.currentExec = self.currentExec + 1
            if (self.executions > 0) and (self.currentExec >= self.executions) then
                self:destroy()
            end
            self.exec(table.unpack(self.arguments))
        end
    end))
    return self
end

function timer.public:unload()
    if not timer.public:isInstance(self) then return false end
    self:destroyInstance()
    return true
end

File: Vital.sandbox/Type/private/timer.cpp
/*----------------------------------------------------------------
     Resource: Vital.sandbox
     Script: Type: public: timer.cpp
     Author: vStudio
     Developer(s): Aviril, Tron, Mario, Аниса, A-Variakojiene
     DOC: 14/09/2022
     Desc: Timer Types
----------------------------------------------------------------*/


//////////////
// Imports //
//////////////

#pragma once
#include <Type/public/timer.h>


/////////////////////////
// Vital: Type: Timer //
/////////////////////////

namespace Vital::Type {
    std::map<Timer*, bool> Timer::buffer;

    // Instantiators //
    Timer::Timer(std::function<void(Timer*)> exec, int interval, int executions) {
        buffer.emplace(this, true);
        Vital::Type::Thread([=](Vital::Type::Thread* thread) -> void {
            int currentExecutions = 0;
            int targetInterval = std::max(0, interval), targetExecutions = std::max(0, executions);
            while (Timer::isInstance(this) && ((targetExecutions == 0) || (currentExecutions < targetExecutions))) {
                thread -> sleep(interval);
                currentExecutions++;
                exec(this);
            };
            destroy();
        }).detach();
    }
    void Timer::destroy() {
        buffer.erase(this);
    }

    // Utils //
    bool Timer::isInstance(Timer* identifier) {
        return buffer.find(identifier) != buffer.end();
    }
}